src/core/observable.ts

```diff
@@ -16,46 +64,115 @@ export interface IObservable extends IDepTreeNode {
 	 * the dependency is already established
 	 */
 	lastAccessedBy: number;
-	staleObservers: IDerivation[];
-	observers: SimpleSet<IDerivation>;
+
+	isPendingUnobservation: boolean; // for effective unobserving
+	isObserved: boolean; // for unobserving only once ber observation
+	// sets of observers grouped their state to only notify about changes.
+	observers: DerivationsSets;
+
 	onBecomeUnobserved();
 }
 
 export function addObserver(observable: IObservable, node: IDerivation) {
 	observable.observers.add(node);
+	if (!observable.isObserved) {
+		observable.isObserved = true;
+	}
 }
 
 export function removeObserver(observable: IObservable, node: IDerivation) {
 	observable.observers.remove(node);
-	if (observable.observers.length === 0)
-		observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
+	if (observable.isObserved && observable.observers.length === 0) {
+		if (globalState.inBatch > 0) {
+			/**
+			 * Wan't to observe/unobserve max once per observable during batch.
+			 * Let's postpone it to end of the batch
+			 */
+			if (!observable.isPendingUnobservation) {
+				observable.isPendingUnobservation = true;
+				globalState.pendingUnobservations.push(observable);
+			}
+		} else {
+			observable.isObserved = false;
+			observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
+		}
+	}
+}
+
+export function startBatch() {
+	globalState.inBatch++;
+}
+
+export function endBatch() {
+	if (--globalState.inBatch === 0) {
+		globalState.pendingUnobservations.splice(0).forEach(observable => {
+			observable.isPendingUnobservation = false;
+			if (observable.isObserved && observable.observers.length === 0) {
+				observable.isObserved = false;
+				observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
```

Comment:
double check, this triggers clearObserving for all cases and results in state -1